src/probnum/kernels/_kernel.py

```diff
@@ -140,135 +364,99 @@ def _check_and_reshape_inputs(
-        output_dim)*. If inputs represent multiple observations, then the resulting
-        matrix has *shape=(n0, n1) or (n0, n1, output_dim, output_dim)*.
 
-        Parameters
-        ----------
-        x0_shape :
-            Shape of the first input to the covariance function.
-        x1_shape :
-            Shape of the second input to the covariance function.
-        """
-        if len(x0_shape) <= 1 and len(x1_shape) <= 1:
-            if self.output_dim == 1:
-                kern_shape = 0
-            else:
-                kern_shape = ()
-        else:
-            kern_shape = (x0_shape[0], x1_shape[0])
-
-        if self.output_dim > 1:
-            kern_shape += (
-                self.output_dim,
-                self.output_dim,
+class IsotropicMixin(abc.ABC):  # pylint: disable=too-few-public-methods
+    r"""Mixin for isotropic kernels.
+
+    An isotropic kernel is a kernel which only depends on the Euclidean norm of the
+    distance between the arguments, i.e.
+
+    .. math ::
+
+        k(x_0, x_1) = k(\lVert x_0 - x_1 \rVert_2).
+
+    Hence, all isotropic kernels are stationary.
+    """
+
+    def _squared_euclidean_distances(
+        self, x0: np.ndarray, x1: Optional[np.ndarray]
+    ) -> np.ndarray:
+        """Implementation of the squared Euclidean distance, which supports kernel
+        broadcasting semantics."""
+        if x1 is None:
+            return np.zeros_like(  # pylint: disable=unexpected-keyword-arg
+                x0,
+                shape=x0.shape[:-1],
             )
 
-        return _utils.as_shape(kern_shape)
+        sqdiffs = (x0 - x1) ** 2
+
+        if sqdiffs.shape[-1] == 1:
+            return self.input_dim * sqdiffs[..., 0]
 
-    @staticmethod
-    def _reshape_kernelmatrix(
-        kerneval: np.ndarray, newshape: ShapeArgType
+        return np.sum(sqdiffs, axis=-1)
+
+    def _euclidean_distances(
+        self, x0: np.ndarray, x1: Optional[np.ndarray]
     ) -> np.ndarray:
-        """Reshape the evaluation of the covariance function.
-
-        Reshape the given evaluation of the covariance function to the correct shape,
-        determined by the inputs x0 and x1. This method is designed to be called by
-        subclasses of :class:`Kernel` in their :meth:`__call__` function to ensure
-        the returned quantity has the correct shape independent of the implementation of
-        the kernel.
-
-        Parameters:
-        -----------
-        kerneval
-            Covariance function evaluated at ``x0`` and ``x1``.
-        newshape :
-            New shape of the evaluation of the covariance function.
-        """
-        if newshape[0] == 0:
-            return _utils.as_numpy_scalar(kerneval.squeeze())
-        else:
-            return kerneval.reshape(newshape)
+        """Implementation of the Euclidean distance, which supports kernel
+        broadcasting semantics."""
+        sqdists = self._squared_euclidean_distances(x0, x1)
+
+        if x1 is None:
```

Comment:
I'd prefer if you return a zero vector here (at the expense of some code duplication) or make a comment to the effect that this is zero. This is not necessarily immediately obvious why this doesn't need a `np.sqrt`.